### PR TITLE
Explicitly set no lifecycle rule in TF for buckets without one (#5561)

### LINKED ISSUE #5561: Explicitly set no lifecycle rule in TF for buckets without one
Followup from #5078 

Specify an S3 bucket lifecycle configuration with an empty rule (`'rule': {}`) for all buckets that do not have a lifecycle rule already configured in the Terraform config.

This should be done dynamically, see `enable_s3_bucket_inventory()` and `block_public_s3_bucket_access()` for inspiration.

@hannes-ucsc: "The ultimate goal behind this is to remove the stale `mandatory_tag_enforcer_lifecycle_s3` lifecycle rules."

The list of affected buckets is here: https://github.com/DataBiosphere/azul/issues/5078#issuecomment-1729993515

----

- [x] Security design review completed; the Resolution of this issue does **not** …
  - [x] … affect authentication; for example:
    - OAuth 2.0 with the application (API or Swagger UI)
    - Authentication of developers with Google Cloud APIs
    - Authentication of developers with AWS APIs
    - Authentication with a GitLab instance in the system
    - Password and 2FA authentication with GitHub
    - API access token authentication with GitHub
    - Authentication with 
  - [x] … affect the permissions of internal users like access to
    - Cloud resources on AWS and GCP
    - GitLab repositories, projects and groups, administration
    - an EC2 instance via SSH
    - GitHub issues, pull requests, commits, commit statuses, wikis, repositories, organizations
  - [x] … affect the permissions of external users like access to
    - TDR snapshots
  - [x] … affect permissions of service or bot accounts
    - Cloud resources on AWS and GCP
  - [x] … affect audit logging in the system, like
    - adding, removing or changing a log message that represents an auditable event
    - changing the routing of log messages through the system
  - [x] … affect monitoring of the system
  - [x] … introduce a new software dependency like
    - Python packages on PYPI
    - Command-line utilities
    - Docker images
    - Terraform providers
  - [x] … add an interface that exposes sensitive or confidential data at the security boundary
  - [x] … affect the encryption of data at rest
  - [x] … require persistence of sensitive or confidential data that might require encryption at rest
  - [x] … require unencrypted transmission of data within the security boundary
  - [x] … affect the network security layer; for example by 
    - modifying, adding or removing firewall rules
    - modifying, adding or removing security groups
    - changing or adding a port a service, proxy or load balancer listens on
- [x] Documentation on any unchecked boxes is provided in comments below

### PR DESCRIPTION
**Pull request by nadove-ucsc**
Connected issues: #5561

### Author

- [x] PR is a draft
- [x] Target branch is `develop`
- [x] Name of PR branch matches `issues/<GitHub handle of author>/<issue#>-<slug>`
- [x] PR title references all connected issues
- [x] PR title matches<sup>1</sup> that of a connected issue <sub>or comment in PR explains why they're different</sub>
- [x] For each connected issue, there is at least one commit whose title references that issue
- [x] PR is connected to all connected issues via ZenHub
- [x] PR description links to connected issues
- [x] Added `p` tag to titles of partial commits
- [x] Added `partial` label to PR <sub>or this PR completely resolves all connected issues</sub>
- [x] All connected issues are resolved partially <sub>or this PR does not have the `partial` label</sub>

<sup>1</sup> when the issue title describes a problem, the corresponding PR
title is `Fix: ` followed by the issue title

### Author (reindex, API changes)

- [x] Added `r` tag to commit title <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR <sub>or this PR does not require reindexing</sub>
- [x] PR and connected issue are labeled `API` <sub>or this PR does not modify a REST API</sub>
- [x] Added `a` (`A`) tag to commit title for backwards (in)compatible changes <sub>or this PR does not modify a REST API</sub>
- [x] Updated REST API version number in `app.py` <sub>or this PR does not modify a REST API</sub>

### Author (chains)

- [x] This PR is blocked by previous PR in the chain <sub>or this PR is not chained to another PR</sub>
- [x] Added `base` label to the blocking PR <sub>or this PR is not chained to another PR</sub>
- [x] Added `chained` label to this PR <sub>or this PR is not chained to another PR</sub>

### Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR <sub>or this PR does not require upgrading</sub>

### Author (operator tasks)

- [x] Added checklist items for additional operator tasks <sub>or this PR does not require additional tasks</sub>

### Author (hotfixes)

- [x] Added `F` tag to main commit title <sub>or this PR does not include permanent fix for a temporary hotfix</sub>
- [x] Reverted the temporary hotfixes for any connected issues <sub>or the `prod` branch has no temporary hotfixes for any connected issues</sub>

### Author (before every review)

- [x] Rebased PR branch on `develop`, squashed old fixups
- [x] Ran `make requirements_update` <sub>or this PR does not touch requirements*.txt, common.mk, Makefile and Dockerfile</sub>
- [x] Added `R` tag to commit title <sub>or this PR does not touch requirements*.txt</sub>
- [x] Added `reqs` label to PR <sub>or this PR does not touch requirements*.txt</sub>
- [x] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>

### Peer reviewer (after requesting changes)

Uncheck the *Author (before every review)* checklists.

### Peer reviewer (after approval)

- [x] PR is not a draft
- [x] Ticket is in *Review requested* column
- [x] Requested review from primary reviewer
- [x] Assigned PR to primary reviewer

### Primary reviewer (after requesting changes)

Uncheck the *before every review* checklists. Update the `N reviews` label.

### Primary reviewer (after approval)

- [x] Actually approved the PR
- [x] Labeled connected issues as `demo` or `no demo`
- [x] Commented on connected issues about demo expectations <sub>or all connected issues are labeled `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [x] `N reviews` label is accurate
- [x] Moved ticket to *Approved* column
- [x] Assigned PR to current operator

### Operator (before pushing merge the commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Checked that demo expectations are clear <sub>or all connected issues are labeled `no demo`</sub>
- [x] PR has checklist items for upgrading instructions <sub>or PR is not labeled `upgrade`</sub>
- [x] Squashed PR branch and rebased onto `develop`
- [x] Sanity-checked history
- [x] Pushed PR branch to GitHub
- [x] Pushed PR branch to GitLab `dev` and added `sandbox` label <sub>or PR is labeled `no sandbox`</sub>
- [x] Pushed PR branch to GitLab `anvildev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Pushed PR branch to GitLab `anvilprod` <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `sandbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `anvilbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `hammerbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build logs for anomalies in `sandbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build logs for anomalies in `anvilbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build logs for anomalies in `hammerbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Deleted unreferenced indices in `sandbox` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Deleted unreferenced indices in `anvilbox` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Deleted unreferenced indices in `hammerbox` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Started reindex in `sandbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Started reindex in `anvilbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Started reindex in `hammerbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `sandbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `anvilbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `hammerbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Title of merge commit starts with title from this PR
- [x] Added PR reference to merge commit title
- [x] Collected commit title tags in merge commit title <sub>but only include `p` if the PR is labeled `partial`</sub>
- [x] Moved connected issues to Merged column in ZenHub
- [x] Pushed merge commit to GitHub

### Operator (chain shortening)

- [x] Changed the target branch of the blocked PR to `develop` <sub>or this PR is not labeled `base`</sub>
- [x] Removed the `chained` label from the blocked PR <sub>or this PR is not labeled `base`</sub>
- [x] Removed the blocking relationship from the blocked PR <sub>or this PR is not labeled `base`</sub>
- [x] Removed the `base` label from this PR <sub>or this PR is not labeled `base`</sub>

### Operator (after pushing the merge commit)

- [x] Deployed `dev.shared`
- [x] Pushed merge commit to GitLab `dev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Deployed `anvildev.shared`
- [x] Pushed merge commit to GitLab `anvildev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Deployed `anvilprod.shared`
- [x] Pushed merge commit to GitLab `anvilprod` <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes on GitLab `dev`<sup>1</sup>
- [x] Reviewed build logs for anomalies on GitLab `dev`<sup>1</sup>
- [x] Build passes on GitLab `anvildev`<sup>1</sup>
- [x] Reviewed build logs for anomalies on GitLab `anvildev`<sup>1</sup>
- [x] Build passes on GitLab `anvilprod`<sup>1</sup>
- [x] Reviewed build logs for anomalies on GitLab `anvilprod`<sup>1</sup>
- [x] Deleted PR branch from GitHub
- [x] Deleted PR branch from GitLab `dev`
- [x] Deleted PR branch from GitLab `anvildev`
- [x] Deleted PR branch from GitLab `anvilprod`

<sup>1</sup> When pushing the merge commit is skipped due to the PR being
labelled `no sandbox`, the next build triggered by a PR whose merge commit *is*
pushed determines this checklist item.

### Operator (reindex)

- [x] Deleted unreferenced indices in `dev` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Deleted unreferenced indices in `anvildev` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Deleted unreferenced indices in `anvilprod` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Started reindex in `dev` <sub>or this PR does not require reindexing</sub>
- [x] Started reindex in `anvildev` <sub>or this PR does not require reindexing</sub>
- [x] Started reindex in `anvilprod` <sub>or this PR does not require reindexing</sub>
- [x] Checked for and triaged indexing failures in `dev` <sub>or this PR does not require reindexing</sub>
- [x] Checked for and triaged indexing failures in `anvildev` <sub>or this PR does not require reindexing</sub>
- [x] Checked for and triaged indexing failures in `anvilprod` <sub>or this PR does not require reindexing</sub>
- [x] Emptied fail queues in `dev` deployment <sub>or this PR does not require reindexing</sub>
- [x] Emptied fail queues in `anvildev` deployment <sub>or this PR does not require reindexing</sub>
- [x] Emptied fail queues in `anvilprod` deployment <sub>or this PR does not require reindexing</sub>

### Operator

- [x] Unassigned PR

## Shorthand for review comments

- `L` line is too long
- `W` line wrapping is wrong
- `Q` bad quotes
- `F` other formatting problem

### CODE REVIEW

**Comment by nadove-ucsc** [comment]
@dsotirho-ucsc Please ignore the CodeQL warning for now.

**Review by dsotirho-ucsc — CHANGES_REQUESTED** [review]
I get a terraform validation failure when trying to build the `shared` component:

```
Creating shared.tf.json
terraform validate
╷
│ Error: Missing required argument
│
│   on shared.tf.json line 254, in resource[7].aws_s3_bucket_lifecycle_configuration[2].trail.rule:
│  254:                         "rule": {}
│
│ The argument "status" is required, but no definition was found.
╵
╷
│ Error: Missing required argument
│
│   on shared.tf.json line 254, in resource[7].aws_s3_bucket_lifecycle_configuration[2].trail.rule:
│  254:                         "rule": {}
│
│ The argument "id" is required, but no definition was found.
╵
╷
│ Error: Missing required argument
│
│   on shared.tf.json line 260, in resource[7].aws_s3_bucket_lifecycle_configuration[3].aws_config.rule:
│  260:                         "rule": {}
│
│ The argument "status" is required, but no definition was found.
╵
╷
│ Error: Missing required argument
│
│   on shared.tf.json line 260, in resource[7].aws_s3_bucket_lifecycle_configuration[3].aws_config.rule:
│  260:                         "rule": {}
│
│ The argument "id" is required, but no definition was found.
╵
make: *** [validate] Error 1
```

**Comment by nadove-ucsc** [comment]
Unfortunately, creating these empty policies does not appear to have the desired effect of removing the existing policies. 
```
$ (cd terraform/shared/ && CI_COMMIT_REF_NAME=develop make plan)
...

  # aws_s3_bucket_lifecycle_configuration.aws_config will be created
  + resource "aws_s3_bucket_lifecycle_configuration" "aws_config" {
      + bucket = "edu-ucsc-gi-platform-hca-dev-awsconfig.us-east-1"
      + id     = (known after apply)

      + rule {
          + id     = "aws_config_default_lifecycle_policy"
          + status = "Enabled"
        }
    }

  # aws_s3_bucket_lifecycle_configuration.trail will be created
  + resource "aws_s3_bucket_lifecycle_configuration" "trail" {
      + bucket = "edu-ucsc-gi-platform-hca-dev-trail.us-east-1"
      + id     = (known after apply)

      + rule {
          + id     = "trail_default_lifecycle_policy"
          + status = "Enabled"
        }
    }

Plan: 2 to add, 0 to change, 0 to destroy.
````

Presumably, Terraform does not consider the rules in the configuration to be incompatible with the preexisting ones, because we are forced to specify an `id` to distinguish them.

**Comment by nadove-ucsc** [comment]
Hannes found an alternative approach that appears to achieve the desired effect at the cost of some deprecation warnings:

```
Terraform will perform the following actions:

  # aws_s3_bucket.aws_config will be updated in-place
  ~ resource "aws_s3_bucket" "aws_config" {
        id                          = "edu-ucsc-gi-platform-hca-dev-awsconfig.us-east-1"
        tags                        = {
            "Name"                = "azul-aws_config"
            "billing"             = "hca"
            "component"           = "azul-aws_config"
            "deployment"          = "dev"
            "owner"               = "hannes@ucsc.edu"
            "service"             = "azul"
            "terraform_component" = "shared"
        }
        # (11 unchanged attributes hidden)

      ~ lifecycle_rule {
          ~ enabled                                = true -> false
            id                                     = "mandatory_tag_enforcer_lifecycle_s3"
            tags                                   = {}
            # (1 unchanged attribute hidden)

          + expiration {
              + days = 36500
            }

          - transition {
              - days          = 1 -> null
              - storage_class = "INTELLIGENT_TIERING" -> null
            }
        }

        # (4 unchanged blocks hidden)
    }

  # aws_s3_bucket.trail will be updated in-place
  ~ resource "aws_s3_bucket" "trail" {
        id                          = "edu-ucsc-gi-platform-hca-dev-trail.us-east-1"
        tags                        = {
            "Name"                = "azul-trail"
            "billing"             = "hca"
            "component"           = "azul-trail"
            "deployment"          = "dev"
            "owner"               = "hannes@ucsc.edu"
            "service"             = "azul"
            "terraform_component" = "shared"
        }
        # (11 unchanged attributes hidden)

      ~ lifecycle_rule {
          ~ enabled                                = true -> false
            id                                     = "mandatory_tag_enforcer_lifecycle_s3"
            tags                                   = {}
            # (1 unchanged attribute hidden)

          + expiration {
              + days = 36500
            }

          - transition {
              - days          = 1 -> null
              - storage_class = "INTELLIGENT_TIERING" -> null
            }
        }

        # (4 unchanged blocks hidden)
    }

Plan: 0 to add, 2 to change, 0 to destroy.

Warnings:

- Argument is deprecated
  on shared.tf.json line 95 (and 3 more)

To see the full warning notes, run Terraform without -compact-warnings.
```

**Review by dsotirho-ucsc — DISMISSED** [review]
Approved.

File: environment.py [hunk 1]
```diff
@@ -605,6 +605,12 @@ def env() -> Mapping[str, Optional[str]]:
         #
         'TF_DATA_DIR': '{project_root}/deployments/.active/.terraform.{AWS_PROFILE}',
 
+        # Make Terraform's deprecation warnings more compact
+        #
+        'TF_CLI_ARGS_validate': '-compact-warnings',
+        'TF_CLI_ARGS_plan': '-compact-warnings',
+        'TF_CLI_ARGS_apply': '-compact-warnings',
```
Comment (hannes-ucsc) [inline environment.py]: Let's `**` a dict comprehension here.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

File: environment.py [hunk 2]
```diff
@@ -605,6 +605,13 @@ def env() -> Mapping[str, Optional[str]]:
         #
         'TF_DATA_DIR': '{project_root}/deployments/.active/.terraform.{AWS_PROFILE}',
 
+        # Make Terraform's deprecation warnings more compact
+        #
+        **{
+            'TF_CLI_ARGS_' + stage: '-compact-warnings'
+            for stage in ['validate', 'plan', 'apply']
```
Comment (hannes-ucsc) [inline environment.py]:
```suggestion
            'TF_CLI_ARGS_' + command: '-compact-warnings'
            for command in ['validate', 'plan', 'apply']
```

"stage" is already taken and `terraform --help` refers to them as "commands".

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

**Review by hannes-ucsc — APPROVED** [review]

**Pull request merged**